File: 0069-sqrtx/0069-sqrtx.py
"""
def shortSolution(x: int) -> int:
    if x == 0 or x == 1:
        return x
            
    low = 0   
    high = x // 2
    while low <= high:
        mid = (low + high) // 2
        res = mid * mid
        
        if res == x:
            return mid
        
        if res < x:
            low = mid + 1
        else:
            high = mid - 1
     
    return high
"""            
def recursiveSolution(x: int, low: int, high: int) -> int:
    mid = (low + high) // 2
    res = mid * mid
    
    # two base cases! 
    
    if res == x:
        return mid

    if res < x and (mid+1)*(mid+1) > x:
        return mid
    
    if res > x:
        return recursiveSolution(x, low, mid-1)
    else:
        return recursiveSolution(x, mid+1, high)
    
    
class Solution:
    def mySqrt(self, x: int) -> int:
        if x == 0 or x == 1:
            return x
        return recursiveSolution(x, 0, x//2)

File: 0069-sqrtx/0069-sqrtx.go
func mySqrt(x int) int {
    if x == 0 || x == 1 {
        return x
    }
    return findRecursively(x, 0, x/2)
}

func findRecursively(x, low, high int) int {
    midpoint := (high + low) / 2
    product := midpoint * midpoint

    if product == x {
        return midpoint
    }

    if product < x && x < (midpoint+1) * (midpoint+1) {
        return midpoint
    }
    
    if product > x {
        return findRecursively(x, low, midpoint-1)
    }
    
    return findRecursively(x, midpoint+1, high)
}

/*
func mySqrt(x int) int {
    if x == 0 || x == 1 {
        return x
    }
    
    low := 0
    high := x / 2
    for low <= high {
        midpoint := (high + low) / 2
        product := midpoint * midpoint
        
        if product == x {
            return midpoint
        }
        
        if product < x && x < (midpoint+1) * (midpoint+1) {
            return midpoint
        }
        
        if product > x {
            high = midpoint - 1
        } else {
            low = midpoint + 1
        }
    }
    
    return high
}
*/